1_ElectronicFieldNotes.py

```diff
@@ -333,6 +335,49 @@ def CheckFVVals(self):
         AquariusUploadManager.CheckFVVals(mode, self)
 
 
+    def UploadToNG(self, server, username, password):
+
+        with timeseries_client(server, username, password) as timeseries:
+            if timeseries.isServerVersionLessThan("14.1"):
+                # Bail if we know the server is 3.X
+                return False, None
+
+            try:
+                locationUniqueId = timeseries.getLocationUniqueId(self.genInfoManager.stnNumCmbo)
+
+                # Save the file to XML, PDF, and zip in a temp folder
+                tempdir = tempfile.mkdtemp()
+
+                xmlName = self.gui.SaveAsXMLAtUpload(tempdir, "")
+                xmlPath = join(tempdir, xmlName)
+                rootPath, _ = os.path.splitext(xmlPath)
+                pdfPath = rootPath + ".pdf"
+                pdfName = os.path.basename(pdfPath)
+                zipPath = rootPath + ".zip"
+                archiveName = join("attachments", pdfName)
+
+                with zipfile.ZipFile(zipPath, 'w') as zip:
+                    zip.write(xmlPath, os.path.basename(xmlPath))
+                    zip.write(pdfPath, archiveName)
+
+                # Upload the ZIP archive
+                with open(zipPath, 'rb') as zipStream:
+                    response = timeseries.acquisition.post("/locations/" + locationUniqueId + "/visits/upload/plugins",
+                                         files={'file': zipStream}).json()
+
+                return True, None
+
+            except requests.exceptions.HTTPError as e:
+
+                return True, str(e) + "\n\nTo show this HTTPError, I had to add 'password' to the response value."
```

Comment:
This hacky bit of code is the main reason I wouldn't merge the code directly.

When an EHSN upload fails, the server responds with a helpful error message that should be shown to the user.

Some common error messages include:
`Saving parsed data would result in duplicates.` - You've already uploaded the file.
`ReferencePoint with name '{name}' for location {id} was not found` - The eHSN file contains level survey activity for benchmarks that don't exist in AQTS. Better add the benchmark and try again.

In order to show this error message, I had to hack in the phrase "password" somewhere in the text. If the response doesn't include the string "password", the existing eHSN upload logic assumes everything is successful. That's quite a logic bug that needs to be fixed, so I hacked in this phrase so that at least the error gets shown.

![image](https://user-images.githubusercontent.com/54773/53829595-5269e900-3f35-11e9-8873-a8b9a9c8dfd8.png)
![image](https://user-images.githubusercontent.com/54773/53829700-95c45780-3f35-11e9-87c1-493c78dbd898.png)

You'll really need to fix the error handling logic throughout the code base to make the eHSN app more robust.